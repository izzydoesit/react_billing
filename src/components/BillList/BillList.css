div.bill-list-wrapper {
  margin: 1%;
  padding: 1% 1% 1% 0;
}

ul.bill-list {
  padding-left: 0;
  line-height: 1.5em;
  list-style: disc inside url(../../bill-icon-bullet.png);
}