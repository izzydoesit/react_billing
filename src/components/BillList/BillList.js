// @flow 

import React, { Component } from 'react';
import Bill from '../Bill/Bill';
import {
  isToday,
  isPastDate,
  isFutureDate,
  formatCentsToDollars } from 'clark-utils';
import type { BillType, BillsType } from '../../types'
import './BillList.css';

type Props = {
  bills?: BillsType,
};

export default class BillList extends Component<Props> {
  static defaultProps = {
    bills: {}
  }

  extractGroup = (groupName: string): Array<BillType> => {
    const bills = this.props.bills;
    let groupList = [];

    if (groupName === 'paid') {
      for (let key in bills) {
        const bill = bills[key];
        if (bill['status'] === 'paid') {
          const billCopy = { ...bill };
          groupList.push(billCopy);
        }
      };
    } else {
      if (groupName === 'outstanding') {
        for (let key in bills) {
          const bill = bills[key];
          if (bill['status'] === 'pending' && (isToday(bill['dueDate']) || isFutureDate(bill['dueDate']))) {
            const billCopy = { ...bill };
            groupList.push(billCopy);
            groupList[groupList.length-1]['status'] = 'outstanding';
          }
        };
      } else if (groupName === 'overdue') {
        for (let key in bills) {
          const bill = bills[key];
          if (bill['status'] === 'pending' && isPastDate(bill['dueDate'])) {
            const billCopy = { ...bill };
            groupList.push(billCopy);
            groupList[groupList.length-1]['status'] = 'overdue';
          }
        };
      }
    }

    return groupList;
  }

  sortGroup = (group: Array<BillType>): Array<BillType> => {
    return group.sort((a, b) => { 
      return (a.dueDate < b.dueDate) ? 1 : ((b.dueDate > a.dueDate) ? -1 : 0);
    });
  }

  render() {
    const paid = this.sortGroup(this.extractGroup('paid'));
    const outstanding = this.sortGroup(this.extractGroup('outstanding'));
    const overdue = this.sortGroup(this.extractGroup('overdue'));
    const sortedBills = [...outstanding, ...overdue, ...paid]

    return (
      <div className="bill-list-wrapper">
        <ul className="bill-list">
          {sortedBills.length === 0
          ? <li className="empty-list-item">No bills have been created...</li>
          : sortedBills.map((bill, index) =>
            <Bill
              key={index}
              id={bill.id}
              status={bill.status}
              amount={formatCentsToDollars(bill.amountInCents)}
              dueDate={bill.dueDate}
            />
          )}
        </ul>
      </div>
    )
  }
}
