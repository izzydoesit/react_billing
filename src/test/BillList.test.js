// @flow

import React from 'react';
import BillList from '../components/BillList/BillList';
import billData from '../data';
import {
  isToday,
  isPastDate,
  isFutureDate, } from 'clark-utils';
import { mount, configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

describe('<BillList />', () => {
  let props, list;
  let mountedBillList;
  const billList = () => {
    if (!mountedBillList) {
      mountedBillList = mount(<BillList {...props} />);
    }
    return mountedBillList;
  };

  beforeEach(() => {
    props = {
      bills: {}
    }
    mountedBillList = undefined;
  });

  it('renders correctly', () => {
    expect(shallowToJson(billList())).toMatchSnapshot();
  });

  it('renders a <div> element', () => {
    expect(billList().find('div.bill-list-wrapper').length).toEqual(1);
  });

  it('renders one <ul> element', () => {
    expect(billList().find('ul.bill-list').length).toEqual(1);
  });

  describe('when bills is defined', () => {
    beforeEach(() => {
      props.bills = billData;
      list = billList().find('ul.bill-list').children();
    });

    it('renders a list of bills', () => {
      const numberOfBills = Object.keys(billData).length;
      expect(list.length).toEqual(numberOfBills);
    });

    it('groups paid bills by status', () => {
      const allPaidBills = Object.keys(billData).filter((key) => {
        const bill = billData[key];
        return bill['status'] === 'paid';
      });
      const lastFourBills = list.slice(-4);
      lastFourBills.forEach((bill) => {
        const billStatus = bill.find('span.bill-status').text();
        expect(billStatus).toContain('paid');
      });
      expect(allPaidBills.length).toEqual(lastFourBills.length);
    });

    it('sorts paid bills by due date', () => {
      const lastFourBills = list.slice(-4);
      let previousBillDueDate = new Date(lastFourBills.first().find('span.bill-due-date').text()).getTime();
      for (let i = 1; i < lastFourBills.length; i++) {
        const currentBillDueDate = new Date(lastFourBills.at(i).find('span.bill-due-date').text()).getTime();
        expect(currentBillDueDate).toBeLessThan(previousBillDueDate);
      }
    });

    it('groups overdue bills by status and due date', () => {
      const allOverdueBills = Object.keys(billData).filter((key) => {
        const bill = billData[key];
        return bill['status'] !== 'paid' && isPastDate(bill['dueDate']);
      });
      const middleThreeBills = list.slice(2, 5);
      middleThreeBills.forEach((bill) => {
        const billStatus = bill.find('span.bill-status').text();
        expect(billStatus).toContain('overdue');
      })
      expect(allOverdueBills.length).toEqual(middleThreeBills.length);
    });

    it('sorts overdue bills by due date', () => {
      const middleThreeBills = list.slice(2, 5);
      let previousBillDueDate = new Date(middleThreeBills.first().find('span.bill-due-date').text()).getTime();
      for (let i = 1; i < middleThreeBills.length; i++) {
        const currentBillDueDate = new Date(middleThreeBills.at(i).find('span.bill-due-date').text()).getTime();
        expect(currentBillDueDate).toBeLessThan(previousBillDueDate);
      }
    });

    it('groups outstanding bills by status and due date', () => {
      const allOutstandingBills = Object.keys(billData).filter((key) => {
        const bill = billData[key];
        return bill['status'] !== 'paid' && (isToday(bill['dueDate']) || isFutureDate(bill['dueDate']));
      });
      const firstTwoBills = list.slice(0, 2);
      firstTwoBills.forEach((bill) => {
        const billStatus = bill.find('span.bill-status').text();
        expect(billStatus).toContain('outstanding');
      });
      expect(allOutstandingBills.length).toEqual(firstTwoBills.length);
    });

    it('sorts outstanding bills by due date', () => {
      const firstTwoBills = list.slice(0, 2);
      let previousBillDueDate = new Date(firstTwoBills.first().find('span.bill-due-date').text()).getTime();
      for (let i = 1; i < firstTwoBills.length; i++) {
        const currentBillDueDate = new Date(firstTwoBills.at(i).find('span.bill-due-date').text()).getTime();
        expect(currentBillDueDate).toBeLessThan(previousBillDueDate);
      }
    });
  });

  describe('when bills is empty', () => {
    beforeEach(() => {
      props.bills = {};
      list = billList().find('ul.bill-list').children();
    });

    it('renders one list item with a "No bills..." message', () => {
      expect(list.length).toEqual(1);
      expect(list.first().text()).toEqual("No bills have been created...")
    });
  });
});
