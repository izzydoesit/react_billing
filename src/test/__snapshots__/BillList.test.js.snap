// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BillList /> renders correctly 1`] = `
<BillList
  bills={Object {}}
>
  <div
    className="bill-list-wrapper"
  >
    <ul
      className="bill-list"
    >
      <li
        className="empty-list-item"
      >
        No bills have been created...
      </li>
    </ul>
  </div>
</BillList>
`;
